rust/ql/lib/codeql/rust/controlflow/internal/Scope.qll

```diff
@@ -0,0 +1,36 @@
+private import rust
+private import Completion
+private import codeql.rust.generated.ParentChild
+
+abstract class CfgScope extends AstNode { }
+
+class FunctionScope extends CfgScope, Function { }
+
+cached
+private module Cached {
+  /**
+   * Gets the immediate parent of a non-`AstNode` element `e`.
+   *
+   * We restrict `e` to be a non-`AstNode` to skip past non-`AstNode` in
+   * the transitive closure computation in `getParentOfAst`. This is
+   * necessary because the parent of an `AstNode` is not necessarily an `AstNode`.
+   */
+  private Element getParentOfAstStep(Element e) {
+    not e instanceof AstNode and
+    result = getImmediateParent(e)
+  }
+
+  /** Gets the nearest enclosing parent of `ast` that is an `AstNode`. */
+  cached
+  AstNode getParentOfAst(AstNode ast) { result = getParentOfAstStep*(getImmediateParent(ast)) }
```

Comment:
This predicate is only used in this file, so I would make it `private` and non-`cached`.

Reply:
And can I then also remove the surrounding `private module`?

Reply:
Yes.